File: Perl/submatrix_selector/submatrix_selector.pl
#! /usr/bin/env perl
use warnings;

#######################################################################
#
# Copyright(c) 2014-2021 Whitehead Institute for Biomedical Research.
#              All Rights Reserved
#
# Given a tab-delimited file matrix, select desired rows and columns.
#
# George Bell, BaRC
# Version 1.1: 2/6/14 	# Removed $dir = "."; [GB]
# Version 1.2: 8/17/21 	# Let program run even if "columns to get" includes items not in matrix [GB]
#
#######################################################################

$exp_data_file = $ARGV[0];
$row_file = $ARGV[1];
$col_file = $ARGV[2];

if (! $exp_data_file && (! ($row_file || $col_file)))
{
	print STDERR "\nSubmatrix selector (select rows and/or columns from a matrix)\n";
	print STDERR "USAGE: $0 matrixFile rowIDfile columnIDfile > submatrixFile[output]\n";
	print STDERR "To select only by row IDs:\n";
	print STDERR "       $0 matrixFile rowIDfile > submatrixFile[output]\n\n";
	print STDERR "To select only by columns IDs:\n";
	print STDERR "       $0 matrixFile 0 columnIDfile > submatrixFile[output]\n\n";
	exit;
}

#####################  analyze  #####################

if ($row_file)	# Only select certain rows 
{
	open  (ROW_LIST, "$row_file") || die "Cannot open ROW_LIST file $row_file: $!";
	while (<ROW_LIST>)
	{
		chomp($_);

		s/^\s+//;
		s/\s+$//;
		s/\r\n/\n/g; # Convert Windows files
		s/\r/\n/g; # Convert Mac files
		
		if ($_)
		{
			push (@rows, $_);
			$rowsToGet{$_} = 1;
		}
	}
	close (ROW_LIST);
}

if ($col_file)	# Only select certain columns 
{
	open  (COL_LIST, "$col_file") || die "Cannot open COL_LIST file $col_file: $!";
	while (<COL_LIST>)
	{
		chomp($_);

		s/^\s+//;
		s/\s+$//;
		s/\r\n/\n/g; # Convert Windows files
		s/\r/\n/g; # Convert Mac files

		if ($_)
		{
			push (@cols, $_);
		}
	}
	close (COL_LIST);
	
	# Get column IDs from matrix file
	open  (EXP_DATA, "$exp_data_file") || die "Cannot open EXP_DATA file $exp_data_file: $!";
	while (<EXP_DATA>)
	{
		if ($. == 1)
		{
			chomp($_);

			s/^\s+//;
			s/\s+$//;
			s/\r\n/\n/g; # Convert Windows files
			s/\r/\n/g; # Convert Mac files

			@columnIDsInMatrix = split (/\t/, $_);
		}
	}
	close(EXP_DATA);
	
	# Added 8/17/2021 to remove columns in "to get" list but not present in matrix
	@cols = get_A_in_B(\@columnIDsInMatrix, \@cols);
}

# Read through big matrix file

open  (EXP_DATA, "$exp_data_file") || die "Cannot open EXP_DATA file $exp_data_file: $!";
$lineNum = 0;
while (<EXP_DATA>)
{
	chomp($_);
	
	s/^\s+//;
	s/\s+$//;
	s/\r\n/\n/g; # Convert Windows files
	s/\r/\n/g; # Convert Mac files

	@fields = split (/\t/, $_);
	
	# To get rid of double quotes added by Excel
	if (/\".+\"/)
	{
		for ($z = 0; $z <= $#fields; $z++)
		{
			if ($fields[$z] =~ /^\".*\"$/)
			{
				$fields[$z] =~ s/^\"(.*)\"$/$1/;
			}
		}
		$_ = join ("\t", @fields);
	}
	
	if (! $lineNum)	# first line; get header info
	{
		if (! $col_file)	# Print all columns
		{
			print "$_\n";
		}
		else	# Print headers for selected columns
		{			
			$colNamesTabbed = join ("\t", @cols);
			
			print $colNamesTabbed;
			print "\n";
		}
	
		# Link header for each column to column number
	
		@headerFields = split (/\t/, $_);
		
		for ($i = 0; $i <= $#headerFields; $i++)
		{
			$colNameToNum{$headerFields[$i]} = $i;
		}
	}
	
	if ($col_file)
	{
		checkChosenColsExist();
	}
	
	# Get this row (either this row is selected or all rows are requested)
	
	if (($fields[0] && $row_file && $rowsToGet{$fields[0]}) || ! $row_file)
	{
		if (! $col_file)	# Get all columns
		{
			if (! $row_file && $lineNum)
			{
				$id2info{$lineNum} = $_;
			}
			else
			{
				$id2info{$fields[0]} = $_;
			}
		}
		else
		{
			# Get the selected columns

			if ($row_file)	# Get only selected rows
			{
				$idKey = $fields[0];
			}
			else	# Get all rows -- and let redundant entries in first column
			{
				$idKey = $lineNum;
			}

			if ($row_file)	# start hash with key in first column
			{
				# Removed 7.6.07
				# $id2info{$idKey} = $fields[0];
			}
			else	# skip first column unless it's requested
			{
				$id2info{$idKey} = "";
			}

			foreach $selectedCol (@cols)
			{
				# Add a trailing tab if there's already something in the hash
				if ($id2info{$idKey})
				{
					$id2info{$idKey} .= "\t";
				}

				if ($colNameToNum{$selectedCol})
				{
					$id2info{$idKey} .= "$fields[$colNameToNum{$selectedCol}]";
				}
				elsif ($colNameToNum{$selectedCol} eq "0")
				{
					$id2info{$idKey} .= "$fields[$colNameToNum{$selectedCol}]";
				}
				else
				{
					$id2info{$idKey} .= "?";
				}
			}
		}
		
		if (! $row_file && $lineNum)	# Get all rows
		{
			push (@rows, $idKey);
			$rowsToGet{$fields[0]} = 1;
		}
	}
	
	$lineNum++;
}
close (EXP_DATA);

if ($row_file)
{
	foreach $inputGene (@rows)
	{
		if ($id2info{$inputGene})
		{
			print "$id2info{$inputGene}\n";
		}
		else
		{
			print "$inputGene\t?\n";
		}
	}
	close (SELECTED_OUT);
}
else	# Print all rows
{
	for ($i = 1; $i < $lineNum; $i++)
	{
		if ($id2info{$i})
		{
			print "$id2info{$i}\n";
		}
		else
		{
			print "$i\tNo data\n";
		}
	}
	close (SELECTED_OUT);
}

#####################  HTML out  #####################

print STDERR "All done.\n\n";

sub checkChosenColsExist
{
	foreach $colPresent (@headerFields)
	{
		$present{$colPresent} = 1;
	}
	foreach $colChosen (@cols)
	{
		if (! $present{$colChosen})
		{
			print STDERR "You've selected a column named \"$colChosen\"\n";
			print STDERR "that doesn't appear to be a column name in your input matrix file\n";
			exit;		
		}
	}
}

sub get_A_in_B
{
	# Added 8/17/2021 to remove columns in "to get" list but not present in matrix
	my ($inMatrix, $toGet) = @_;	
	my @inMatrix = @{ $inMatrix };
	my @toGet = @{ $toGet };
	
	%inMatrix = ();
	@toGetInMatrix = ();
	
	foreach $item (@inMatrix)
	{
		$inMatrix{$item} = 1;
	}

	foreach $item (@toGet)
	{
		if ($inMatrix{$item})
		{
			push @toGetInMatrix, $item;
			# print STDERR "Found column \"$item\" (present in matrix).\n";
		}
		else
		{
			print STDERR "Skipping column \"$item\" (not present in matrix).\n";
		}
	}
	return @toGetInMatrix;
}

File: Python/submatrix_selector/submatrix_selector.py
#! /usr/bin/env python3

#######################################################################
#
# Copyright(c) 2022 Whitehead Institute for Biomedical Research.
#			  All Rights Reserved
#
# Given a tab-delimited file matrix, select desired rows and columns.
# Adapted from submatrix_selector.pl
# Version 1.0: 7 March 2022 
#
# George Bell, BaRC
#
#######################################################################

import sys
import csv

# Ensure that at least two arguments were entered
try:
	matrixFile = sys.argv[1]
	rowFile = sys.argv[2]
except IndexError:
	sys.stderr.write("\nSubmatrix selector (select rows and/or columns from a matrix)\n")
	sys.stderr.write("USAGE: submatrix_selector.py matrixFile rowIDfile columnIDfile > submatrixFile[output]\n\n")
	sys.stderr.write("To select only by row IDs:\n")
	sys.stderr.write("       submatrix_selector.py matrixFile rowIDfile > submatrixFile[output]\n\n")
	sys.stderr.write("To select only by columns IDs:\n")
	sys.stderr.write("       submatrix_selector.py matrixFile 0 columnIDfile > submatrixFile[output]\n\n")
	exit()
# Get this (optional) argument
try:
	colFile = sys.argv[3]
except IndexError:
	sys.stderr.write("No column file given.  Printing all columns ....\n")
	colFile = "0"

if (rowFile == "0"):
	sys.stderr.write("No row file given.  Printing all rows ....\n")

# Initialize lists of rows and columns we want
rowsToGet = []
colsToGet = []

# sys.stderr.write("Col file: " + colFile)

# Read the list of row IDs to print
if (rowFile and rowFile != "0"):
	with open(rowFile) as rows:
		reader = csv.reader(rows, delimiter="\t")
		for row in reader:
			# print(row)
			rowsToGet.append(row[0])
row_dictionary = dict.fromkeys(rowsToGet, "get")

# Read the list of column IDs to print
if (colFile and colFile != "0"):
	# sys.stderr.write("Reading " + colFile + " to get columns to print ....\n")
	with open(colFile) as cols:
		reader = csv.reader(cols, delimiter="\t")
		for col in reader:
			colsToGet.append(col[0])
col_dictionary = dict.fromkeys(colsToGet, "get")

# Read through big matrix file
with open(matrixFile) as matrixData:
	reader = csv.reader(matrixData, delimiter="\t")
	for matrixRow in reader:
		fieldsToPrint = []
		if (reader.line_num == 1):
			matrixHeaderRow = matrixRow
			for item in matrixHeaderRow:
				if item in col_dictionary.keys() or colFile == "0":
					fieldsToPrint.append(item)
			print("\t".join(fieldsToPrint))
		else:
			# Iterate through this row (and header row, again)
			for headerItem, item in zip(matrixHeaderRow, matrixRow):
				if headerItem in col_dictionary.keys() or colFile == "0":
					fieldsToPrint.append(item)
			# Is the first field of this row in our "rows to get"?
			if matrixRow[0] in row_dictionary.keys() or rowFile == "0":
				print("\t".join(fieldsToPrint))

# End
